serial.c

```diff
@@ -452,6 +490,235 @@ static void Interrupt com_general_isr(void)
+    unsigned long baseaddr;
+    __dpmi_meminfo region;
+
+    if(__dpmi_get_segment_base_address(_my_ds(), &baseaddr) == 0)
+    {
+        region.handle = 0;
+        region.size = sizeof(g_comports);
+        region.address = baseaddr + (unsigned long)g_comports;
+        __dpmi_unlock_linear_region(&region);
+    }
+    if(__dpmi_get_segment_base_address(_my_cs(), &baseaddr) == 0)
+    {
+        region.handle = 0;
+        region.size = ISR_SIZE;
+        region.address = baseaddr + (unsigned long)com_general_isr;
+        __dpmi_unlock_linear_region(&region);
+    }
+    _go32_dpmi_free_iret_wrapper(&g_isr_addr);
+}
+
+#else /* ! __DJGPP__ */
+
+static void serial_dpmi_get_pvect(int vect, int_handler_ptr *handler)
+{
+    union REGS r;
+    unsigned short sel;
+    unsigned long off;
+
+    /* DPMI get protected mode interrupt vector: Int 31H, Fn 0204H */
+    r.x.eax = 0x0204;
+    r.x.ebx = vect;
+    int386(0x31, &r, &r);
+    sel = (unsigned short) r.x.ecx;
+    off = r.x.edx;
+
+    *handler=(int_handler_ptr) MK_FP(sel, off);
+}
+
+static void serial_dpmi_set_pvect(int vect, int_handler_ptr *handler)
+{
+    union REGS r;
+    void Far *ptr;
+
+    /* DPMI set protected mode interrupt vector: Int 31H, Fn 0205H */
+    ptr = (void Far *)*handler;
+    r.x.eax = 0x0205;
+    r.x.ebx = vect;
+    r.x.ecx = FP_SEG(ptr);
+    r.x.edx = FP_OFF(ptr);
+    int386(0x31, &r, &r);
+}
+
+static int serial_dpmi_lock_linear_memory(void Far *ptr, unsigned long size)
+{
+    union REGS r;
+
+    /* DPMI get segment base address: Int 31H, Fn 0006H */
+    r.x.eax = 0x0006;
+    r.x.ebx = FP_SEG(ptr);
+    int386(0x31, &r, &r);
+    if(r.w.cflag == 0)
+    {
+        unsigned long addr = FP_OFF( ptr ) + ((r.w.cx << 16) | r.w.dx);
+
+        /* DPMI lock linear region: Int 31H, Fn 0600H */
+        r.x.eax = 0x0600;
+        r.x.ebx = addr >> 16;
+        r.x.ecx = addr & 0xFFFF;
+        r.x.esi = size >> 16;
+        r.x.edi = size & 0xFFFF;
+        int386(0x31, &r, &r);
+        if(r.w.cflag == 0)
+        {
+            g_isr_addr = com_general_isr;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static void serial_dpmi_unlock_linear_memory(void Far *ptr, unsigned long size)
+{
+    union REGS r;
+
+    /* DPMI get segment base address: Int 31H, Fn 0006H */
+    r.x.eax = 0x0006;
+    r.x.ebx = FP_SEG(ptr);
+    int386(0x31, &r, &r);
+    if(r.w.cflag == 0)
+    {
+        unsigned long addr = FP_OFF( ptr ) + ((r.w.cx << 16) | r.w.dx);
+
+        /* DPMI unlock linear region: Int 31H, Fn 0601H */
+        r.x.eax = 0x0601;
+        r.x.ebx = addr >> 16;
+        r.x.ecx = addr & 0xFFFF;
+        r.x.esi = size >> 16;
+        r.x.edi = size & 0xFFFF;
+        int386(0x31, &r, &r);
+    }
+    int386 (0x31, &r, &r);
+}
+
+static int serial_dpmi_lock_memory(void)
+{
+    if(serial_dpmi_lock_linear_memory(com_general_isr, ISR_SIZE))
+    {
+        if(serial_dpmi_lock_linear_memory(g_comports, sizeof(g_comports)))
+            return 1;

```

Comment:
Minor quibble, but I always use curly braces even on single liners because it's bit me sooooo many times when refactoring code.